spec/lib/requests_rate_limiter_spec.rb

```diff
@@ -0,0 +1,280 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+describe RequestsRateLimiter do
+  def assert_rate_limit_response(response, expected)
+    status, headers, body = response
+    expect(status).to eq(429)
+    expect(headers["Discourse-Rate-Limit-Error-Code"]).to eq(expected)
+    expect(body.first).to include("Error code: #{expected}")
+  end
+
+  class TestLogger
+    attr_accessor :warnings
+
+    def initialize
+      @warnings = 0
+    end
+
+    def warn(*args)
+      @warnings += 1
+    end
+  end
+
+  fab!(:user) { Fabricate(:user, trust_level: 2) }
+
+  def instance(env = {}, u = nil)
+    RequestsRateLimiter.new(
+      user_id: u&.id,
+      trust_level: u&.trust_level,
+      request: Rack::Request.new(env)
+    )
+  end
+
+  before do
+    @old_logger = Rails.logger
+    Rails.logger = TestLogger.new
+    RateLimiter.enable
+    RateLimiter.clear_all!
+    RateLimiter.clear_all_global!
+  end
+
+  after do
+    Rails.logger = @old_logger
+    RateLimiter.disable
+  end
+
+  describe "#skip_limits?" do
+    it "returns true if rate limit mode is neither block nor warn" do
+      global_setting :max_reqs_per_ip_mode, ""
+      expect(instance.skip_limits?). to eq(true)
+
+      global_setting :max_reqs_per_ip_mode, "nothing"
+      expect(instance.skip_limits?). to eq(true)
+    end
+
+    it "returns false if rate limit mode is either block or warn or both" do
+      global_setting :max_reqs_per_ip_mode, "block"
+      expect(instance.skip_limits?). to eq(false)
+
+      global_setting :max_reqs_per_ip_mode, "warn"
+      expect(instance.skip_limits?). to eq(false)
+
+      global_setting :max_reqs_per_ip_mode, "warn+block"
+      expect(instance.skip_limits?). to eq(false)
+    end
+
+    it "returns true if the request ip is private and private ips are excluded from rate limits" do
+      global_setting :max_reqs_rate_limit_on_private, false
+      expect(instance({ "REMOTE_ADDR" => "192.168.1.1" }).skip_limits?).to eq(true)
+      expect(instance({ "REMOTE_ADDR" => "193.178.1.10" }).skip_limits?).to eq(false)
+      global_setting :max_reqs_rate_limit_on_private, true
+      expect(instance({ "REMOTE_ADDR" => "192.168.1.1" }).skip_limits?).to eq(false)
+    end
+
+    it "returns true if ip_skipper callback returns true" do
+      orig_callback = Middleware::RequestTracker.ip_skipper
+      Middleware::RequestTracker.unregister_ip_skipper
+      Middleware::RequestTracker.register_ip_skipper do |ip|
+        ip == "193.211.211.4"
+      end
+      expect(instance({ "REMOTE_ADDR" => "103.211.211.4" }).skip_limits?).to eq(false)
+      expect(instance({ "REMOTE_ADDR" => "193.211.211.4" }).skip_limits?).to eq(true)
+    ensure
+      Middleware::RequestTracker.unregister_ip_skipper
+      Middleware::RequestTracker.register_ip_skipper(orig_callback) if orig_callback
+    end
+  end
+
+  describe "#apply_limits!" do
```

Comment:
I think we need to include tests to demonstrate how the rate limit does not apply to users that has the required trust level.

Reply:
(rate limits are still applied to trusted users, the difference is they have their own limits that are not shared with anyone else like I explained yesterday). The tests that start with `returns the right error response when applying...` in this file and `request_tracker_spec.rb` test/demonstrate this behavior.